crates/node/src/shims/abcipp_shim.rs

```diff
@@ -114,43 +113,89 @@ impl AbcippShim {
                 Req::EndBlock(_) => {
                     let begin_block_request =
                         self.begin_block_request.take().unwrap();
-                    let block_time = begin_block_request
-                        .header
-                        .time
-                        .try_into()
-                        .expect("valid RFC3339 block time");
 
-                    let tm_raw_hash_string = tm_raw_hash_to_string(
-                        begin_block_request.header.proposer_address,
-                    );
-                    let block_proposer = find_validator_by_raw_hash(
-                        &self.service.state,
-                        tm_raw_hash_string,
-                    )
-                    .unwrap()
-                    .expect(
-                        "Unable to find native validator address of block \
-                         proposer from tendermint raw hash",
-                    );
+                    let (process_proposal_result, processing_results) =
+                        match namada::core::hash::Hash::try_from(
+                            begin_block_request.hash,
+                        ) {
+                            Ok(block_hash) => {
+                                match self
+                                    .service
+                                    .state
+                                    .in_mem()
+                                    .process_proposal_cache
+                                    .get(&block_hash)
+                                {
+                                    Some((process_resp, res)) => (
+                                        // We already have the result of
+                                        // process proposal for this block
+                                        // cached in memory
+                                        process_resp.to_owned(),
+                                        res.iter()
+                                            .map(|res| res.to_owned().into())
+                                            .collect(),
+                                    ),
+                                    None => {
```

Comment:
so, as far as I can tell, the only reason why a process proposal wouldn't be in cache is because you shut down your node, clearing it from memory, right? under normal circumstances, you always receive a `ProcessProposal` request before you receive an `EndBlock` request. under these assumptions, you would never/seldom hit the `None` branch

Reply:
I believe there's one case in which a node could receive an `EndBlock` call without having the result cached. It's explained here in the comet docs https://docs.cometbft.com/v0.37/spec/abci/abci++_example_scenarios#scenario-3

Reply:
ah right, if the system is ahead of you